### PR TITLE
Added support for sql restriction params, projection aliases and specifying the max results retreived

### PR DESCRIPTION
**Pull request by Adonaioc**

### CODE REVIEW

**Comment by bdw429s** [comment]
@Adonaioc I needed to make a fix for the permissions on the Travis build.  Can you please rebase your branch against `development` to get those commits.  Then we can see if your pull passes.

**Comment by bdw429s** [comment]
@Adonaioc Actually, I was able to fix the issue with the build related to permissions.  It looks like the unit tests are now legitimately failing.  Can you please check why the tests are failing?  You can view the output here:
https://travis-ci.org/coldbox-modules/cbox-cborm/builds/170812077

File: modules/cborm/models/criterion/Restrictions.cfc [hunk 1]
```diff
@@ -169,8 +169,21 @@ component singleton{
 		return restrictions.sizeNE(arguments.property, arguments.propertyValue);
 	}
 	
-	// Use arbitrary SQL to modify the resultset
-	any function sqlRestriction(required string sql){
+	any function sqlRestriction(required string sql, any parameters ){
+        //if we defined a parameter then its assume its a string and pass it through right
```
Comment (lmajano) [inline modules/cborm/models/criterion/Restrictions.cfc]: Could there be any other types in stead of just string? What about numeric, floats, etc?

File: modules/cborm/models/DetachedCriteriaBuilder.cfc [hunk 2]
```diff
@@ -75,7 +75,16 @@ component accessors="true" extends="cborm.models.BaseBuilder" {
...
 		// now that we have the sql string, we can create the sqlProjection
-		return this.PROJECTIONS.sqlProjection( sql, [ alias ], SQLHelper.getProjectedTypes() );
+		var projection =  this.PROJECTIONS.sqlProjection( sql, [ alias ], SQLHelper.getProjectedTypes() );
+        // finally, add the alias to the projection list so we can sort on the column if needed
+        return this.PROJECTIONS.alias( projection, alias );
+
+	}
+
+	// Set a limit upon the number of objects to be retrieved.
+    any function maxResults(required numeric maxResults){
```
Comment (lmajano) [inline modules/cborm/models/DetachedCriteriaBuilder.cfc]: This is separate from the sql restrictions update right?

File: modules/cborm/models/criterion/Restrictions.cfc (hunk 1 above)
Comment (Adonaioc) [inline modules/cborm/models/criterion/Restrictions.cfc]: yes, I just havent had a chance to add them, I will try to get around to it in week or so.

File: modules/cborm/models/DetachedCriteriaBuilder.cfc (hunk 2 above)
Comment (Adonaioc) [inline modules/cborm/models/DetachedCriteriaBuilder.cfc]: yes its an added function that allows you to limit the result set

**Comment by bdw429s** [comment]
Any updates on this?

**Comment by lmajano** [comment]
I think we need some tests around the functionality.

**Comment by Adonaioc** [comment]
So I have to admit I have not used the unit testing capabilities of your coldbox project before and am having a bit of trouble getting it to run locally, do I need to install something else to make it work?

**Comment by bdw429s** [comment]
Have you followed the instructions here?
https://github.com/ColdBox/coldbox-platform/blob/development/tests/readme.md

File: modules/cborm/models/criterion/Restrictions.cfc (hunk 1 above)
Comment (Adonaioc) [inline modules/cborm/models/criterion/Restrictions.cfc]: There are many many different types that can be defined so should I make it default to string if not passed in but allow the programmer to define it if they like?

**Comment by Adonaioc** [comment]
I got the tests running locally and it looks like only one is failing and its because I changed the return type of createDetachedSQLProjection from sqlProjection to org.hibernate.criterion.AliasedProjection which implements EnhancedProjection which extends projection. The reason for this is to allow sorting on detached criteria, Should I modify the test to check the new type or create a separate function for this with a new test or something entirely different?

**Comment by jclausen** [comment]
@Adonaioc : I see this PR is hanging out there.  I would suggest just modifying the return type check on the test, unless @lmajano has any objections.

File: modules/cborm/models/criterion/Restrictions.cfc (hunk 1 above)
Comment (lmajano) [reply modules/cborm/models/criterion/Restrictions.cfc]: I have merged this manually into be, but have added full support for all types.

**Comment by Adonaioc** [comment]
Thanks! I no longer work with cf or coldbox but I love what you are doing in the community!

File: modules/cborm/models/BaseBuilder.cfc [hunk 3]
```diff
@@ -557,8 +557,8 @@ component accessors="true"{
 		var partialSQL = "";
 		projection.sql = "";
 		// if multiple subqueries have been specified, smartly separate them out into a sql string that will work
-		for( var x=1; x<=listLen( arguments.rawProjection.sql ); x++ ) {
-			partialSQL = listGetAt( arguments.rawProjection.sql, x );
+		for( var x=1; x<=listLen( arguments.rawProjection.sql, "&&" ); x++ ) {
```
Comment (lmajano) [inline modules/cborm/models/BaseBuilder.cfc]: I am not sure I follow why the adding of the `&&` can you refresh my memory why?

**Comment by Adonaioc** [comment]
It might have been part of something for another modification to the module that was not fully included in the pull request. We extended the orm module quite a bit. I no longer have access to the codebase but I don’t see how it would be applicable by itself in this context.

**Comment by lmajano** [comment]
Thanks so much @Adonaioc , I will be releasing a major update to the module this week. Thanks for the feedback and too bad you are not using ColdBox anymore. Don't forget about us and send us features or ideas 👍 

https://github.com/coldbox-modules/cborm/blob/development/changelog.md

**Pull request closed without merging**